File: Solution Depth-First Search.js

/**
 * @param {string[]} equations
 * @return {boolean}
 */
var equationsPossible = function (equations) {
    this.ALPHABET_SIZE = 26;
    this.asciiCodeFor_a = 97;
    this.ROOT_NOT_MARKED = -1;

    this.rootsOfEqualElements = new Array(this.ALPHABET_SIZE).fill(this.ROOT_NOT_MARKED);
    this.graph = new Array(this.ALPHABET_SIZE);
    initializeGraph(equations);

    for (let i = 0; i < this.ALPHABET_SIZE; ++i) {
        if (this.rootsOfEqualElements[i] === this.ROOT_NOT_MARKED) {
            markRootByDepthFirstSearch(i, i);
        }
    }

    for (let equation of equations) {
        if (equation.charAt(1) === '!' && rootsAreEqual(equation)) {
            return false;
        }
    }
    return true;
};

/**
 * @param {number} element
 * @param {number} root
 * @return {void}
 */
function markRootByDepthFirstSearch(element, root) {
    if (this.rootsOfEqualElements[element] !== this.ROOT_NOT_MARKED) {
        return;
    }
    this.rootsOfEqualElements[element] = root;
    for (let current of this.graph[element]) {
        markRootByDepthFirstSearch(current, root);
    }
}

/**
 * @param {string[]} equations
 * @return {void}
 */
function initializeGraph(equations) {
    for (let i = 0; i < this.ALPHABET_SIZE; ++i) {
        this.graph[i] = new Set();
    }

    for (let equation of equations) {
        if (equation.charAt(1) === '=') {
            this.graph[equation.codePointAt(0) - this.asciiCodeFor_a].add(equation.codePointAt(3) - this.asciiCodeFor_a);
            this.graph[equation.codePointAt(3) - this.asciiCodeFor_a].add(equation.codePointAt(0) - this.asciiCodeFor_a);
        }
    }
}

/**
 * @param {string} equation
 * @return {boolean}
 */
function rootsAreEqual(equation) {
    return this.rootsOfEqualElements[equation.codePointAt(0) - this.asciiCodeFor_a] === this.rootsOfEqualElements[equation.codePointAt(3) - this.asciiCodeFor_a];
}

File: Solution Depth-First Search.java

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class Solution {

    private static final int ALPHABET_SIZE = 26;
    private static final int ROOT_NOT_MARKED = -1;

    int[] rootsOfEqualElements = new int[ALPHABET_SIZE];
    Set<Integer>[] graph = new HashSet[ALPHABET_SIZE];

    public boolean equationsPossible(String[] equations) {
        Arrays.fill(rootsOfEqualElements, ROOT_NOT_MARKED);
        initializeGraph(equations);

        for (int i = 0; i < ALPHABET_SIZE; ++i) {
            if (this.rootsOfEqualElements[i] == ROOT_NOT_MARKED) {
                markRootByDepthFirstSearch(i, i);
            }
        }

        for (String equation : equations) {
            if (equation.charAt(1) == '!' && rootsAreEqual(equation)) {
                return false;
            }
        }
        return true;
    }

    private void markRootByDepthFirstSearch(int element, int root) {
        if (this.rootsOfEqualElements[element] != ROOT_NOT_MARKED) {
            return;
        }
        this.rootsOfEqualElements[element] = root;
        for (int current : graph[element]) {
            markRootByDepthFirstSearch(current, root);
        }
    }

    private void initializeGraph(String[] equations) {
        for (int i = 0; i < ALPHABET_SIZE; ++i) {
            graph[i] = new HashSet<>();
        }

        for (String equation : equations) {
            if (equation.charAt(1) == '=') {
                graph[equation.charAt(0) - 'a'].add(equation.charAt(3) - 'a');
                graph[equation.charAt(3) - 'a'].add(equation.charAt(0) - 'a');
            }
        }
    }

    private boolean rootsAreEqual(String equation) {
        return rootsOfEqualElements[equation.charAt(0) - 'a'] == rootsOfEqualElements[equation.charAt(3) - 'a'];
    }
}
